File: src/code_intelligence/detectors/shell/bash_detector.py
"""Bash/Shell script detector for functions, sourced files, exports, and tool usage."""

from __future__ import annotations

import re

from code_intelligence.detectors.base import DetectorContext, DetectorResult
from code_intelligence.models.graph import (
    EdgeKind,
    GraphEdge,
    GraphNode,
    NodeKind,
    SourceLocation,
)

_FUNC_RE = re.compile(r'(?:function\s+(\w+)|(\w+)\s*\(\s*\))\s*\{')
_SOURCE_RE = re.compile(r'(?:source|\.) (?:")?([^\s"]+)')
_SHEBANG_RE = re.compile(r'^#!\s*/(?:usr/)?(?:bin/)?(?:env\s+)?(\w+)')
_EXPORT_RE = re.compile(r'export\s+(\w+)=')

_INFRA_TOOLS = {"docker", "kubectl", "terraform", "aws", "az", "gcloud"}
_TOOL_RE = re.compile(
    r'\b(' + '|'.join(re.escape(t) for t in sorted(_INFRA_TOOLS)) + r')\b'
)


class BashDetector:
    """Detects Bash/Shell script structures: functions, sourced files, exports, and tool usage."""

    name: str = "bash"
    supported_languages: tuple[str, ...] = ("bash",)

    def detect(self, ctx: DetectorContext) -> DetectorResult:
        result = DetectorResult()
        text = ctx.content.decode("utf-8", errors="replace")
        lines = text.split("\n")

        # Detect shebang
        if lines:
            m = _SHEBANG_RE.match(lines[0])
            if m:
                shell = m.group(1)
                result.nodes.append(GraphNode(
                    id=ctx.file_path,
                    kind=NodeKind.MODULE,
                    label=ctx.file_path,
                    fqn=ctx.file_path,
                    module=ctx.module_name,
                    location=SourceLocation(
                        file_path=ctx.file_path,
                        line_start=1,
                    ),
                    properties={"shell": shell},
                ))

        # Detect function definitions
        for i, line in enumerate(lines):
            m = _FUNC_RE.search(line)
            if not m:
                continue
            func_name = m.group(1) or m.group(2)

            result.nodes.append(GraphNode(
                id=f"{ctx.file_path}:{func_name}",
                kind=NodeKind.METHOD,
                label=func_name,
                fqn=func_name,
                module=ctx.module_name,
                location=SourceLocation(
                    file_path=ctx.file_path,
                    line_start=i + 1,
                ),
            ))

        # Detect source/dot-source imports
        for i, line in enumerate(lines):
            m = _SOURCE_RE.search(line)
            if not m:
                continue
            sourced_file = m.group(1)
            result.edges.append(GraphEdge(
                source=ctx.file_path,
                target=sourced_file,
                kind=EdgeKind.IMPORTS,
                label=f"{ctx.file_path} sources {sourced_file}",
            ))

        # Detect exports
        for i, line in enumerate(lines):
            m = _EXPORT_RE.search(line)
            if not m:
                continue
            var_name = m.group(1)

            result.nodes.append(GraphNode(
                id=f"{ctx.file_path}:export:{var_name}",
                kind=NodeKind.CONFIG_DEFINITION,
                label=f"export {var_name}",
                fqn=var_name,
                module=ctx.module_name,
                location=SourceLocation(
                    file_path=ctx.file_path,
                    line_start=i + 1,
                ),
            ))

        # Detect infrastructure tool usage
        tools_seen: set[str] = set()
        for i, line in enumerate(lines):
            # Skip comments
            stripped = line.lstrip()
            if stripped.startswith('#'):
                continue
            for m in _TOOL_RE.finditer(line):
                tool = m.group(1)
                if tool not in tools_seen:
                    tools_seen.add(tool)
                    result.edges.append(GraphEdge(
                        source=ctx.file_path,
                        target=tool,
                        kind=EdgeKind.CALLS,
                        label=f"{ctx.file_path} uses {tool}",
                        properties={"tool": tool},
                    ))

        return result

File: src/main/java/io/github/randomcodespace/iq/detector/script/shell/BashDetector.java
package io.github.randomcodespace.iq.detector.script.shell;

import io.github.randomcodespace.iq.detector.AbstractRegexDetector;
import io.github.randomcodespace.iq.detector.DetectorContext;
import io.github.randomcodespace.iq.detector.DetectorResult;
import io.github.randomcodespace.iq.model.CodeEdge;
import io.github.randomcodespace.iq.model.CodeNode;
import io.github.randomcodespace.iq.model.EdgeKind;
import io.github.randomcodespace.iq.model.NodeKind;
import org.springframework.stereotype.Component;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import io.github.randomcodespace.iq.detector.DetectorInfo;

@DetectorInfo(
    name = "bash",
    category = "config",
    description = "Detects Bash script structure (functions, source imports, variables)",
    languages = {"bash"},
    nodeKinds = {NodeKind.CONFIG_DEFINITION, NodeKind.METHOD, NodeKind.MODULE},
    edgeKinds = {EdgeKind.CALLS, EdgeKind.IMPORTS}
)
@Component
public class BashDetector extends AbstractRegexDetector {

    private static final Pattern FUNC_RE = Pattern.compile("(?:function\\s+(\\w+)|(\\w+)\\s*\\(\\s*\\))\\s*\\{");
    private static final Pattern SOURCE_RE = Pattern.compile("(?:source|\\.) (?:\")?([^\\s\"]+)");
    private static final Pattern SHEBANG_RE = Pattern.compile("^#!\\s*/(?:usr/)?(?:bin/)?(?:env\\s+)?(\\w+)");
    private static final Pattern EXPORT_RE = Pattern.compile("export\\s+(\\w+)=");
    private static final Pattern TOOL_RE = Pattern.compile("\\b(aws|az|docker|gcloud|kubectl|terraform)\\b");

    @Override
    public String getName() { return "bash"; }

    @Override
    public Set<String> getSupportedLanguages() { return Set.of("bash"); }

    @Override
    public DetectorResult detect(DetectorContext ctx) {
        String text = ctx.content();
        if (text == null || text.isEmpty()) return DetectorResult.empty();

        List<CodeNode> nodes = new ArrayList<>();
        List<CodeEdge> edges = new ArrayList<>();
        String fp = ctx.filePath();
        String[] lines = text.split("\n", -1);

        if (lines.length > 0) {
            Matcher m = SHEBANG_RE.matcher(lines[0]);
            if (m.find()) {
                CodeNode n = new CodeNode(); n.setId(fp);
                n.setKind(NodeKind.MODULE); n.setLabel(fp); n.setFqn(fp);
                n.setFilePath(fp); n.setLineStart(1);
                n.getProperties().put("shell", m.group(1));
                nodes.add(n);
            }
        }

        for (int i = 0; i < lines.length; i++) {
            Matcher m = FUNC_RE.matcher(lines[i]);
            if (m.find()) {
                String funcName = m.group(1) != null ? m.group(1) : m.group(2);
                CodeNode n = new CodeNode(); n.setId(fp + ":" + funcName);
                n.setKind(NodeKind.METHOD); n.setLabel(funcName); n.setFqn(funcName);
                n.setFilePath(fp); n.setLineStart(i + 1);
                nodes.add(n);
            }

            m = SOURCE_RE.matcher(lines[i]);
            if (m.find()) {
                CodeEdge e = new CodeEdge(); e.setId(fp + ":sources:" + m.group(1));
                e.setKind(EdgeKind.IMPORTS); e.setSourceId(fp);
                e.setTarget(new CodeNode(m.group(1), NodeKind.MODULE, m.group(1)));
                edges.add(e);
            }

            m = EXPORT_RE.matcher(lines[i]);
            if (m.find()) {
                String varName = m.group(1);
                CodeNode n = new CodeNode(); n.setId(fp + ":export:" + varName);
                n.setKind(NodeKind.CONFIG_DEFINITION); n.setLabel("export " + varName);
                n.setFqn(varName); n.setFilePath(fp); n.setLineStart(i + 1);
                nodes.add(n);
            }
        }

        Set<String> toolsSeen = new HashSet<>();
        for (int i = 0; i < lines.length; i++) {
            String stripped = lines[i].stripLeading();
            if (stripped.startsWith("#")) continue;
            Matcher m = TOOL_RE.matcher(lines[i]);
            while (m.find()) {
                String tool = m.group(1);
                if (toolsSeen.add(tool)) {
                    CodeEdge e = new CodeEdge(); e.setId(fp + ":calls:" + tool);
                    e.setKind(EdgeKind.CALLS); e.setSourceId(fp);
                    e.setTarget(new CodeNode(tool, NodeKind.MODULE, tool));
                    e.getProperties().put("tool", tool);
                    edges.add(e);
                }
            }
        }

        return DetectorResult.of(nodes, edges);
    }
}
